File: 백준/Gold/3217. malloc/malloc.py
import sys
input = sys.stdin.readline

class Node:
    def __init__(self, start, size):
        self.start = start
        self.size = size
        self.next = None

class MemoryManager:
    def __init__(self, total_size):
        self.free_list = Node(1, total_size)
        self.vars = {}

    def allocate(self, var_name, size):
        prev = None
        curr = self.free_list

        while curr:
            if curr.size >= size:
                alloc_start = curr.start
                curr.start += size
                curr.size -= size

                if curr.size == 0:
                    if prev:
                        prev.next = curr.next
                    else:
                        self.free_list = curr.next

                self.vars[var_name] = (alloc_start, size)
                return alloc_start
            prev = curr
            curr = curr.next

        self.vars[var_name] = 0
        return 0

    def deallocate(self, var_name):
        if var_name not in self.vars or self.vars[var_name] == 0:
            return

        start, size = self.vars[var_name]
        new_node = Node(start, size)
        self.vars[var_name] = 0

        if not self.free_list or self.free_list.start > start:
            new_node.next = self.free_list
            self.free_list = new_node
        else:
            curr = self.free_list
            while curr.next and curr.next.start < start:
                curr = curr.next
            new_node.next = curr.next
            curr.next = new_node

        self.merge_free_blocks()

    def merge_free_blocks(self):
        curr = self.free_list
        while curr and curr.next:
            if curr.start + curr.size == curr.next.start:
                curr.size += curr.next.size
                curr.next = curr.next.next
            else:
                curr = curr.next

    def print_address(self, var_name):
        if var_name not in self.vars or self.vars[var_name] == 0:
            print(0)
        else:
            print(self.vars[var_name][0])


mem_manager = MemoryManager(100000)

for _ in range(int(input().strip())):
    line = input().strip()
    
    if '=' in line:
        var_name, cmd = line.split('=')
        size = int(cmd.split('(')[-1].rstrip(');'))
        mem_manager.allocate(var_name.strip(), size)
    elif line.startswith('print'):
        var_name = line.split('(')[-1].rstrip(');')
        mem_manager.print_address(var_name)
    elif line.startswith('free'):
        var_name = line.split('(')[-1].rstrip(');')
        mem_manager.deallocate(var_name)

File: 백준/Gold/3217. malloc/malloc.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MEMORY_SIZE 100000
#define VAR_SIZE 1000

typedef struct Node {
    int start;
    int size;
    struct Node* next;
} Node;

Node* free_list = NULL;
int var_table[VAR_SIZE]; 
int var_size_table[VAR_SIZE]; 
char var_names[VAR_SIZE][5]; 
int var_count = 0; 

void init_memory() {
    free_list = (Node*)malloc(sizeof(Node));
    free_list->start = 1;
    free_list->size = MEMORY_SIZE;
    free_list->next = NULL;
}

int find_var_index(const char* var_name) {
    for (int i = 0; i < var_count; i++) {
        if (strcmp(var_names[i], var_name) == 0) {
            return i;
        }
    }
    strcpy(var_names[var_count], var_name);
    var_table[var_count] = 0;
    var_size_table[var_count] = 0;
    return var_count++;
}

int malloc_memory(int size) {
    Node* current = free_list;
    Node* prev = NULL;

    while (current) {
        if (current->size >= size) {
            int alloc_start = current->start;
            current->start += size;
            current->size -= size;

            if (current->size == 0) {
                if (prev) {
                    prev->next = current->next;
                } else {
                    free_list = current->next;
                }
                free(current);
            }

            return alloc_start;
        }
        prev = current;
        current = current->next;
    }

    return 0;
}

void free_memory(int start, int size) {
    Node* new_node = (Node*)malloc(sizeof(Node));
    new_node->start = start;
    new_node->size = size;
    new_node->next = NULL;

    if (!free_list || free_list->start > start) {
        new_node->next = free_list;
        free_list = new_node;
    } else {
        Node* current = free_list;
        while (current->next && current->next->start < start) {
            current = current->next;
        }
        new_node->next = current->next;
        current->next = new_node;
    }

    Node* current = free_list;
    while (current && current->next) {
        if (current->start + current->size == current->next->start) {
            Node* to_free = current->next;
            current->size += current->next->size;
            current->next = current->next->next;
            free(to_free);
        } else {
            current = current->next;
        }
    }
}

int main() {
    int N;
    scanf("%d", &N);
    char command[30];

    init_memory();

    for (int i = 0; i < N; i++) {
        scanf("%s", command);
        if (strstr(command, "malloc")) {
            char var_name[5];
            int size;
            sscanf(command, "%4s=malloc(%d);", var_name, &size);
            int var_index = find_var_index(var_name);
            int addr = malloc_memory(size);
            var_table[var_index] = addr;
            var_size_table[var_index] = addr ? size : 0; 
        } else if (strstr(command, "free")) {
            char var_name[5];
            sscanf(command, "free(%4s);", var_name);
            int var_index = find_var_index(var_name);
            if (var_table[var_index]) {
                free_memory(var_table[var_index], var_size_table[var_index]);
                var_table[var_index] = 0;
                var_size_table[var_index] = 0;
            }
        } else if (strstr(command, "print")) {
            char var_name[5];
            sscanf(command, "print(%4s);", var_name);
            int var_index = find_var_index(var_name);
            printf("%d\n", var_table[var_index]);
        }
    }

    return 0;
}
